File: recursion/quick_sort.js
var swap = function(array, firstIndex, secondIndex) {
    var temp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = temp;
};

var partition = function(array, p, r) {
    var q = p;
    for (var j = p; j < r; j++) {
        if (array[j] <= array[r]) {
            swap(array, q, j);
            q++;
        }
    }
    swap(array, r, q);
    return q;
};

var quickSort = function(array, p, r) {
  if (p < r) {
    var index = partition(array, p, r);
    quickSort(array, p, index - 1);
    quickSort(array, index + 1, r);
  }
}

var array = [5, 2, 8, 16, 12];
quickSort(array, 0, 4);
console.log(array);

File: recursion/quick_sort.rb
def swap(array, first, second)
  value = array[first]
  array[first] = array[second]
  array[second] = value
end

def partition(array, p, r)
  q = p
  i = p
  while i < r
    if array[i] <= array[r]
      swap(array, q, i)
      q += 1
    end
    i += 1
  end
  swap(array, r, q)
  return q
end

def quick_sort(array, p, r)
  if p < r
    index = partition(array, p, r)
    quick_sort(array, p, index - 1)
    quick_sort(array, index + 1, r)
  end
end

array = [5, 2, 8, 16, 12]
quick_sort(array, 0, 4)
p array
